src/ess_rhat.jl

```diff
@@ -233,25 +236,206 @@ The ESS and ``\\widehat{R}`` values can be computed for the following estimators
 - `StatsBase.mad`
 - `Base.Fix2(Statistics.quantile, p::Real)`
 
+## Types
+
+If no `estimator` is provided, the following `type`s of ESS estimates may be computed:
+- `:bulk`/`:rank`: mean-ESS computed on rank-normalized draws. This type diagnoses poor
+    convergence in the bulk of the distribution due to trends or different locations of the
+    chains.
+- `:tail`: minimum of the quantile-ESS for the symmetric quantiles where
+    `tail_prob=0.1` is the probability in the tails. This type diagnoses poor convergence in
+    the tails of the distribution. If this type is chosen, `kwargs` may contain a
+    `tail_prob` keyword.
+- `:basic`: basic ESS, equivalent to specifying `estimator=Statistics.mean`.
+
+While Bulk-ESS is conceptually related to basic ESS, it is well-defined even if the chains
+do not have finite variance.[^VehtariGelman2021]. For each parameter, rank-normalization
+proceeds by first ranking the inputs using "tied ranking" and then transforming the ranks to
+normal quantiles so that the result is standard normally distributed. This transform is
+monotonic.
+
 [^VehtariGelman2021]: Vehtari, A., Gelman, A., Simpson, D., Carpenter, B., & Bürkner, P. C. (2021).
     Rank-normalization, folding, and localization: An improved ``\\widehat {R}`` for
     assessing convergence of MCMC. Bayesian Analysis.
     doi: [10.1214/20-BA1221](https://doi.org/10.1214/20-BA1221)
     arXiv: [1903.08008](https://arxiv.org/abs/1903.08008)
 """
-function ess_rhat(samples::AbstractArray{<:Union{Missing,Real},3}; kwargs...)
-    return ess_rhat(Statistics.mean, samples; kwargs...)
+@constprop :aggressive function ess(
```

Comment:
Why is `@constprop` needed?

Reply:
If we don't have it, then functions like `foo(x) = ess(x; type=:bulk)` are not type-inferrable.

Reply:
Sure, but do you know why? Shouldn't this be handled by constant propagation? Is this a bug/deficiency of the Julia compiler that should be fixed? Or do we have some code pattern that prevents constant propagation and could be fixed to help the compiler?

Reply:
I don't fully understand why default const-propagation doesn't work here. I'd asked about it on Julia Slack (https://julialang.slack.com/archives/C6A044SQH/p1677318248072949), and this was the recommended solution. An alternative would be to use ValSplit.jl, but that seemed to be a more complex solution.

Reply:
Type inference seems to work if you branch explicitly instead of using `Val`: https://julialang.slack.com/archives/C6A044SQH/p1677365245386389?thread_ts=1677318248.072949&cid=C6A044SQH

Reply:
I adopted this approach.